src/v2/manifest/mod.rs

```diff
@@ -77,22 +89,59 @@ impl Client {
                 }
             })
             .and_then(|res| {
-                future::ok(res.headers().clone()).join(
+                future::ok((res.headers().clone(), res.url().clone())).join(
                     res.into_body()
                         .concat2()
                         .map_err(|e| Error::from(format!("{}", e))),
                 )
             })
             .map_err(|e| Error::from(format!("{}", e)))
-            .and_then(|(headers, body)| {
+            .and_then(|((headers, url), body)| {
                 let content_digest = match headers.get("docker-content-digest") {
                     Some(content_digest_value) => Some(content_digest_value.to_str()?.to_string()),
                     None => {
                         debug!("cannot find manifestref in headers");
                         None
                     }
                 };
-                Ok((body.to_vec(), content_digest))
+
+                let header_content_type = headers.get(header::CONTENT_TYPE);
+                let media_type = evaluate_media_type(header_content_type, &url)?;
+
+                trace!(
+                    "content-type: {:?}, media-type: {:?}",
+                    header_content_type,
+                    media_type
+                );
+
+                Ok((body, content_digest, media_type))
+            })
+            .and_then(move |(body, content_digest, media_type)| {
+                match media_type {
+                    mediatypes::MediaTypes::ManifestV2S1Signed => A(futures::future::result(
+                        serde_json::from_slice::<ManifestSchema1Signed>(&body)
+                            .map_err(|e| Error::from(format!("{}", e)))
+                            .map(Manifest::S1Signed),
+                    )),
+                    mediatypes::MediaTypes::ManifestV2S2 => B(futures::future::result(
+                        serde_json::from_slice::<ManifestSchema2Spec>(&body),
+                    )
+                    .map_err(|e| Error::from(format!("{}", e)))
+                    .and_then(move |m| {
+                        m.fetch_config_blob(client_spare0, name.to_string())
+                            .map(Manifest::S2)
+                    })),
+                    mediatypes::MediaTypes::ManifestList => A(futures::future::result(
+                        serde_json::from_slice::<ManifestList>(&body)
+                            .map_err(|e| Error::from(format!("{}", e)))
+                            .map(Manifest::ML),
+                    )),
+                    unsupported => A(futures::future::result(Err(Error::from(format!(
```

Comment:
You can directly use `future::result` here.

Reply:
Ah, I think you meant `future::err`, right? Thanks in any case for the tip

Reply:
Oh, even better :+1: